From 6a39866dcfc65fb44cfea13f9455b71a516b0f4f Mon Sep 17 00:00:00 2001
From: Muna Sinada <quic_msinada@quicinc.com>
Date: Wed, 2 Mar 2022 15:29:58 -0800
Subject: [PATCH 2/6] ru_puncturing: add configuration option

- New option 'ru_punct_bitmap' to configure RU puncturing bitmap.
- New option 'ru_punct_ofdma' which indicates if kernel should
  consider OFDMA pattern.

Signed-off-by: Muna Sinada <quic_msinada@quicinc.com>
Signed-off-by: Aloka Dixit <quic_alokad@quicinc.com>
---
 hostapd/config_file.c |  4 ++++
 hostapd/hostapd.conf  | 11 +++++++++++
 src/ap/ap_config.h    |  2 ++
 3 files changed, 17 insertions(+)

--- a/hostapd/hostapd.conf
+++ b/hostapd/hostapd.conf
@@ -1107,6 +1107,17 @@ wmm_ac_vo_acm=0
 # will be used as the AP MLD MAC address.
 #mld_addr=02:03:04:05:06:07
 
+# RU puncturing bitmap (16 bits) where each bit corresponds to
+# a 20 MHz channel in the given bandwidth, bit 0 corresponding to the channel
+# with lowest frequency.
+# Bit set to 1 indicates that the channel is punctured, otherwise active.
+# Default value is 0 indicating that no channel is punctured.
+#ru_punct_bitmap=4
+
+# If the kernel should consider OFDMA patterns while validating the provided
+# RU puncturing bitmap. Default value is 0.
+#ru_punct_ofdma=1
+
 ##### IEEE 802.1X-2004 related configuration ##################################
 
 # Require IEEE 802.1X authorization
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -1242,6 +1242,8 @@ struct hostapd_config {
 #define CH_SWITCH_EHT_ENABLED BIT(0)
 #define CH_SWITCH_EHT_DISABLED BIT(1)
 	unsigned int ch_switch_eht_config;
+	u16 ru_punct_bitmap;
+	u8 ru_punct_ofdma;
 
 	enum mbssid {
 		MBSSID_DISABLED = 0,
--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -5044,6 +5044,10 @@ static int hostapd_config_fill(struct ho
 		conf->eht_phy_capab.su_beamformee = atoi(pos);
 	} else if (os_strcmp(buf, "eht_mu_beamformer") == 0) {
 		conf->eht_phy_capab.mu_beamformer = atoi(pos);
+	} else if (os_strcmp(buf, "ru_punct_bitmap") == 0) {
+		conf->ru_punct_bitmap = atoi(pos);
+	} else if (os_strcmp(buf, "ru_punct_ofdma") == 0) {
+		conf->ru_punct_ofdma = atoi(pos);
 	} else if (os_strcmp(buf, "eht_default_pe_duration") == 0) {
 		conf->eht_default_pe_duration = atoi(pos);
 	} else if (os_strcmp(buf, "punct_bitmap") == 0) {
